cpp/include/cudf/detail/groupby/sort_helper.hpp

```diff
@@ -0,0 +1,221 @@
+   * rows in `keys` will not contribute to any aggregation. 
+   *
+   * @param keys table to group by
+   * @param ignore_null_keys ignore rows in keys with nulls
+   * @param null_sort_behavior whether to put nulls before valid values or after
+   * @param keys_pre_sorted if the keys are already sorted
+   * @param stream used for all the computation in this helper object
+   */
+  helper(table_view const& keys, bool ignore_null_keys = true,
+          std::vector<null_order> null_sort_order = {},
+          bool keys_pre_sorted = false)
+  : _keys(keys)
+  , _num_keys(-1)
+  , _ignore_null_keys(ignore_null_keys)
+  , _keys_pre_sorted(keys_pre_sorted)
+  {
+    if (keys_pre_sorted and
+        ignore_null_keys and
+        has_nulls(keys)
+       )
+    {
+      _keys_pre_sorted = false;
+    }
+  };
+
+  ~helper() = default;
+  helper(helper const&) = delete;
+  helper& operator=(helper const&) = delete;
+  helper(helper&&) = default;
+  helper& operator=(helper&&) = default;
+
+  /**
+   * @brief Groups a column of values according to `keys` and sorts within each group.
+   * 
+   * Groups the @p values where the groups are dictated by key table
+   * and each group is sorted in ascending order, with NULL elements positioned 
+   * at the end of each group.
+   * 
+   * @throws cudf::logic_error if `values.size != keys.num_rows()`
+   * 
+   * @param values The value column to group and sort
+   * @return the sorted and grouped column
+   */
+  std::unique_ptr<column> sorted_values(column_view const& values, 
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+    cudaStream_t stream = 0);
+
+  // TODO (dm): implement
+  std::unique_ptr<column> grouped_values(column_view const& values, 
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+    cudaStream_t stream = 0);
+
+  /**
+   * @brief Get a table of sorted unique keys
+   * 
+   * @return a new table in which each row is a unique row in the sorted key table.
+   */
+  std::unique_ptr<table> unique_keys(
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+    cudaStream_t stream = 0);
+
+  /**
+   * @brief Get the number of groups in `keys`
+   * 
+   */
+  size_type num_groups() { return group_offsets().size(); }
+
+  /**
+   * @brief Return the effective number of keys
+   * 
+   * When ignore_null_keys = false, returned value is same as `keys.num_rows()`
+   * When ignore_null_keys = true, returned value is the number of rows in `keys`
+   *  in which no element is null
+   */
+  size_type num_keys();
+
+  /**
+   * @brief Get the sorted order of `keys`.
+   *
+   * Gathering `keys` by sort order indices will produce the sorted key table.
+   * 
+   * Computes and stores the key sorted order on first invocation, and returns the
+   * stored order on subsequent calls.
+   * 
+   * @return the sort order indices for `keys`.
+   */
+  column_view key_sort_order(cudaStream_t stream = 0);
```

Comment:
`key_sort_order` is for internal use. This wraps around the `sorted_order` API which can be used externally. That's why not `unique_ptr` and no memory resource. Also it is memoised using a unique pointer member of sort_helper. Otherwise we'd have to recalculate it everytime

Reply:
I realized later that you are creating and using same element, should have removed this comment .